django/utils/decorators.py

```diff
@@ -17,13 +17,29 @@ def __get__(self, instance, owner):
         return super(classonlymethod, self).__get__(instance, owner)
 
 
-def method_decorator(decorator):
+def method_decorator(decorator, name=""):
     """
     Converts a function decorator into a method decorator
     """
     # 'func' is a function at the time it is passed to _dec, but will eventually
     # be a method of the class it is defined on.
-    def _dec(func):
+    # 'name' is required when a class is decorated instead of a method,
+    # and is the name of the method to be decorated, e.g. "dispatch".
+    def _dec(obj):
+        is_class = isinstance(obj, type)
+        if is_class:
+            if name and hasattr(obj, name):
+                func = getattr(obj, name)
```

Comment:
Going back to the two cases. We can consider 
```python
func = getattr(obj, name, None)
if name and not callable(func):
    raise ValueError("The keyword...")
```
as being "optimistic". In fact, by using `gettattr()` without first checking for the attribute, we make the assumption that the decorator will be used on classes. Note that if the attribute is not found, getting the default makes `getattr()` a relatively expensive operation:

```python
obj = ()

def deco(obj, name):
    func = getattr(obj, name, None)
    if name and not callable(func):
        raise ValueError("The keyword...")
```

If we do some simple profiling of the two possible usages:

```
>>> # When method_decorator() is used on classes:
>>> print(timeit.timeit("deco(obj, 'count')", setup=optimistic))
0.227811098099
>>> # When used on methods (old behavior):
>>> print(timeit.timeit("deco(obj, '')", setup=optimistic))
0.449029922485
```
If we use a pessimistic approach instead, by using `hasattr()`, which introduces a relatively inexpensive check:

```python
def deco(obj, name):
    if name and hasattr(obj, name):
        func = getattr(obj, name)
        if not callable(func):
            raise TypeError("Cannot decorate")
```
Let's test this:
```
>>> # When method_decorator() is used on classes:
>>> print(timeit.timeit("deco(obj, 'count')", setup=pessimistic))
0.341710090637
```
*slightly* more expensive, as predicted. However:
```
>>> # When used on methods (old behavior):
>>> print("Pessimistic, method: " + str(timeit.timeit("deco(obj, '')", setup=pessimistic)))
0.0912799835205
```
in this case the hit is negligible, in the context of a web application.

Since we are introducing a new feature, and we expect that most of the existing code base of an application will be using the old way, it pays off to be pessimistic.